Core/ExpectedException/ExpectedExceptionRewriter.cs

```diff
@@ -0,0 +1,80 @@
+// to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using NUnit2To3SyntaxConverter.ExpectedException.Validators;
+using NUnit2To3SyntaxConverter.Validation;
+using Serilog;
+
+namespace NUnit2To3SyntaxConverter.ExpectedException
+{
+  public class ExpectedExceptionRewriter : CSharpSyntaxRewriter
+  {
+    private readonly ExpectedExceptionAttributeRemover _attributeRemover;
+    private readonly ExpectedExceptionMethodBodyTransformer _methodBodyTransformer;
+    private readonly SemanticModel _semanticModel;
+    private readonly IValidator<MethodDeclarationSyntax> _validator;
+
+    public ExpectedExceptionRewriter (
+        SemanticModel semanticModel,
+        ExpectedExceptionMethodBodyTransformer methodBodyTransformer,
+        ExpectedExceptionAttributeRemover attributeRemover)
+    {
+      _semanticModel = semanticModel;
+      _methodBodyTransformer = methodBodyTransformer;
+      _attributeRemover = attributeRemover;
+      _validator = new ExpectedExceptionValidator();
+    }
+
+    public override SyntaxNode VisitMethodDeclaration (MethodDeclarationSyntax node)
+    {
+      var errors = _validator.Validate (node).ToList();
+
+      if (errors.Count > 0)
+      {
+        foreach (var error in errors)
+        {
+          Log.Warning ("{@file} - {@method}: {@category}:\n {@message}", error.FileName, error.MethodName, error.Category, error.Reason);
+        }
+
+        return node;
+      }
+
+      var expectedExceptionAttribute = QueryExpectedExceptionAttributes (node).SingleOrDefault();
+
+      if (expectedExceptionAttribute == null) return node;
+
+      var withTransformedBody = _methodBodyTransformer.Transform (node, expectedExceptionAttribute);
+      return _attributeRemover.Transform (withTransformedBody, expectedExceptionAttribute);
+    }
+
+    private IEnumerable<ExpectedExceptionModel> QueryExpectedExceptionAttributes (BaseMethodDeclarationSyntax node)

```

Comment:
should this method return a sequence or a single value? there can only ever be a single expectedexpcetionattribute and getting a sequence + doing a single-check outside splits this logic into two parts.